src/main/java/com/amazon/opendistroforelasticsearch/security/resolver/IndexResolverReplacer.java

```diff
@@ -208,49 +209,41 @@ private Resolved resolveIndexPatterns(final IndicesOptions indicesOptions, final
             remoteIndices = Collections.emptySet();
         }
 
-        final Set<String> matchingAliases;
-        final Set<String> matchingIndices;
-        final Set<String> matchingAllIndices;
+        final Collection<String> matchingAliases;
+        final Collection<String> matchingAllIndices;
 
         if (isLocalAll(requestedPatterns0)) {
             if (log.isTraceEnabled()) {
                 log.trace(Arrays.toString(requestedPatterns0) + " is an LOCAL ALL pattern");
             }
             matchingAliases = Resolved.All_SET;
-            matchingIndices = Resolved.All_SET;
             matchingAllIndices = Resolved.All_SET;
 
         } else if (!remoteIndices.isEmpty() && localRequestedPatterns.isEmpty()) {
             if (log.isTraceEnabled()) {
                 log.trace(Arrays.toString(requestedPatterns0) + " is an LOCAL EMPTY request");
             }
-            return new Resolved.Builder().addOriginalRequested(Arrays.asList(requestedPatterns0)).addRemoteIndices(remoteIndices).build();
+            return new Resolved(Collections.EMPTY_SET, Collections.EMPTY_SET, Collections.EMPTY_SET, Arrays.asList(requestedPatterns0), remoteIndices);
         }
 
         else {
 
             ClusterState state = clusterService.state();
 
             final SortedMap<String, AliasOrIndex> lookup = state.metaData().getAliasAndIndexLookup();
-            final Set<String> aliases = lookup.entrySet().stream().filter(e -> e.getValue().isAlias()).map(e -> e.getKey())
-                    .collect(Collectors.toSet());
-
-            matchingAliases = new HashSet<>(localRequestedPatterns.size() * 10);
-            matchingIndices = new HashSet<>(localRequestedPatterns.size() * 10);
-            matchingAllIndices = new HashSet<>(localRequestedPatterns.size() * 10);
+            final List<String> aliases = lookup.entrySet().stream().filter(e -> e.getValue().isAlias()).map(e -> e.getKey())
```

Comment:
I don't understand `Set` to `List` conversion. Please explain why is it necessary and what is the significance of duplicates in the `List`?

Reply:
There are no duplicates in aliases for the simple reason it originates as Set.

The problem is collecting from a Set to Set create a new Set which is significantly more expansive then collecting to a list (or array for that matter).

Use of set may stand for 2 things:

1. We want to clarify in our interface that no duplicates are to be expected and (typically) that order of elements is not insignificant.
2. We want to remove duplicates from original source so we create a Set.

Even while the 1st is generally fine it sometimes has to stay just List and with clarifying documentation about uniqueness of items in performance sensitive code.

This is the case of 1st that got extended to internal and even scope-local places where it serves no purpose except for defensive programming. And internal state is a poor place for such practices, if need be we'd use assertions.
It's not without cost as each time we re-create a Set from another set we'd got to add all elements to a new set, recomputing hash codes (in case of HashSet) or rebalancing the Tree (in case of TreeSet).

Reply:
For this particular `Set` -> `List` conversion, the source is indeed `Set`, but in many other places the source is not a `Set` and, I guess, is subject to duplicates.

What is the expected size of those sets/lists? It is not data, it is security metadata and I would not expect it to exceed 100K.  With 100K the difference between creating `Set` and `List` is negligible.

Reply:
The numbers above speak for themselves I think.

Also the ratio of time to build a hash set vs time to build a list is never close to 1. HashSet is always ~X times more expansive especially with strings as list building will not even look at characters but hash set has to compute a hash for each of them, possibly multiple times due to rehash.

The absolute value may be fairly low, but this operation here is done many thousands of times. Building a million of hash sets of size 10 is way more expansive then lists of size 10. As I already did the profiling and presented the numbers, I suggest you to do the same and see if it indeed will not make a difference

Reply:
I do not want to say that collecting to `Set` is faster than collecting to `List`, but 

- `Set` provides additional functionality and it is not clear to me that the functionality of a `Set` is not required in all cases where `Set` was replaced with a `List`.
- The difference between collecting 1M times of an array of size 10 that contains strings with the length of 100 to `Set` vs `List` is negligible (200ms vs 150ms on my laptop).

Reply:
As far as cases go - that is the precise problem with curent defensively programmed code - I had to reverse engineer what really needs to be a set and what not. If missed something feel free to point out. 

First I want to warn you of danger of doing correct microbenchmarks esp on Java, a platform with powerful JIT compiler featuring a number of agressive dynamic optimizations.
Thankfully for Java there is JMH, a tool built specifically for this purpose and used by JVM engineers as well.

Here is a complete benchmark:
```
import org.openjdk.jmh.annotations.Benchmark;
import org.openjdk.jmh.annotations.Param;
import org.openjdk.jmh.annotations.Scope;
import org.openjdk.jmh.annotations.State;

import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;
import java.util.stream.Stream;


@State(Scope.Benchmark)
public class MyBenchmark {

    public Set<String> data = Stream.iterate(0, x -> x +1)
            .limit(10).map(x -> "logs-some-system"+x)
            .collect(Collectors.toSet());

    @Benchmark
    public List<String> collectToList() {
        return data.stream().filter(x -> x.startsWith("logs")).collect(Collectors.toList());
    }

    @Benchmark
    public Set<String> collectToSet() {
        return data.stream().filter(x -> x.startsWith("logs")).collect(Collectors.toSet());
    }
}
```

And the output on my laptop, sadly the numbers for sets have bigger deal of error margin maybe I was running something else on the laptop, I will re-run it later to check again.
```
REMEMBER: The numbers below are just data. To gain reusable insights, you need to follow up on
why the numbers are the way they are. Use profilers (see -prof, -lprof), design factorial
experiments, perform baseline and negative tests that provide experimental control, make sure
the benchmarking environment is safe on JVM/OS/HW level, ask for reviews from the domain experts.
Do not assume the numbers tell you what you want them to tell.

Benchmark                   Mode  Cnt        Score        Error  Units
MyBenchmark.collectToList  thrpt   25  5873007.810 ±  90055.803  ops/s
MyBenchmark.collectToSet   thrpt   25  3838232.740 ± 468729.614  ops/s

```

This gives me rough estimate  it is collecting a set of 10 short strings with simple filter to set is 50% slower  then to list. Asymptotic for larger sets are interesting too.

Reply:
Made strings a bit longer and 100 items per set. 
Here are the numbers:
```
Benchmark                   Mode  Cnt       Score      Error  Units
MyBenchmark.collectToList  thrpt   50  597804.200 ± 9090.985  ops/s
MyBenchmark.collectToSet   thrpt   50  240782.714 ± 3631.795  ops/s
```
Now sets are ~0.40 times as fast as lists, or conversely list ~2.5x faster.

P.S. I forgot that Strings in Java cache the hashCode so the effect of hashing is dimnished in the benchmark (because it's the same literals over and over) but in Opendistro with HashSets left and right the strings likely to have hashCode cached by that time as well.

Reply:
My point is not that there are cases where `Set` is required and cases where `Collection` or `List` is sufficient. The point is that the time spent in collecting to `List` or `Set` is a small percentage compared to the time spent in other operations and changing `Set` to `Collection` is not well justified even if collecting to `Set` takes 2x or 3x more time collecting to `List` or any other collection.

Sure, we can go more scientific, correct and accurate route and use JMH (I am quite familiar with JMH, hot spot JVM compiler, JIT and vectorization), but it  produces similar (I am interested in the order of magnitude) results:

```
package com.amazon.opendistroforelasticsearch;

import org.openjdk.jmh.annotations.Benchmark;
import org.openjdk.jmh.annotations.BenchmarkMode;
import org.openjdk.jmh.annotations.Measurement;
import org.openjdk.jmh.annotations.Mode;
import org.openjdk.jmh.annotations.OutputTimeUnit;
import org.openjdk.jmh.annotations.Param;
import org.openjdk.jmh.annotations.Scope;
import org.openjdk.jmh.annotations.Setup;
import org.openjdk.jmh.annotations.State;
import org.openjdk.jmh.annotations.Warmup;

import java.nio.charset.Charset;
import java.util.Arrays;
import java.util.HashSet;
import java.util.Random;
import java.util.Set;
import java.util.concurrent.TimeUnit;
import java.util.stream.Collectors;

@State(Scope.Benchmark)
@BenchmarkMode(Mode.AverageTime)
@OutputTimeUnit(TimeUnit.NANOSECONDS)
@Warmup(iterations = 5, time = 100, timeUnit = TimeUnit.MILLISECONDS)
@Measurement(iterations = 5, time = 100, timeUnit =  TimeUnit.MILLISECONDS)
public class CollectorsBenchmark {

    @Param({"10"})
    private int SIZE;

    @Param({"100"})
    private int LENGTH;

    private String[] array;
    private Set<String> set;
    private final Charset charset = Charset.forName("UTF-8");

    @Setup
    public void setup() {
        Random random = new Random();
        array = new String[SIZE];
        set = new HashSet<>(SIZE);
        for (int i = 0; i < SIZE; i++) {
            byte[] string = new byte[LENGTH];
            random.nextBytes(string);
            array[i] = new String(string, charset);
            set.add(array[i]);
        }
    }

    @Benchmark
    public void benchmarkSetToList() {
        set.stream().collect(Collectors.toList());
    }

    @Benchmark
    public void benchmarkSetToSet() {
        set.stream().collect(Collectors.toSet());
    }

    @Benchmark
    public void benchmarkArrayToList() {
        Arrays.stream(array).collect(Collectors.toList());
    }

    @Benchmark
    public void benchmarkArrayToSet() {
        Arrays.stream(array).collect(Collectors.toSet());
    }
}
```
```
Benchmark                                 (LENGTH)  (SIZE)  Mode  Cnt    Score    Error  Units
CollectorsBenchmark.benchmarkArrayToList       100      10  avgt   25   63.526 ±  5.563  ns/op
CollectorsBenchmark.benchmarkArrayToSet        100      10  avgt   25  149.283 ± 16.372  ns/op
CollectorsBenchmark.benchmarkSetToList         100      10  avgt   25  121.944 ± 21.390  ns/op
CollectorsBenchmark.benchmarkSetToSet          100      10  avgt   25  193.206 ± 18.709  ns/op
```

In this particular case, `String` is a key of the `lookup` (`SortedMap`) so its hash code is computed and cached by the time keys are collected to a `Set` or a `List`.

Reply:
> The point is that the time spent in collecting to List or Set is a small percentage compared to the time spent in other operations and changing Set to Collection is not well justified even if collecting to Set takes 2x or 3x more time collecting to List or any other collection.

This is quite interesting as I recall you prompted me to avoid boxing a boolean and to use collect to array instead of collect to List. Was the reason puristic following of rules or was “prefer more efficient version” your motivation? 

In this particular case (object is local to this function) the semantic gain of having Set instead of List is ZERO but it costs more. If you absolutely want it - I’ll put it back.

Most of other places are also either internal helpers for this file or the callsite just iterates the elements. 

Please point out specific places where you believe having Set is semantically important to justify the extra cost in performance and I’ll roll them back.

Reply:
The difference is that boxing a boolean or collecting to a `List` instead of an array does not provide any additional functionality compared to collecting to a `List` instead of a `Set`. The same applies to the usage of `ConcurrentHashMap` where `HashMap` is sufficient. It is necessary to guarantee the correctness of data first and then see how it can be obtained more efficiently. In addition, using `List` instead of a `Set` may be OK for now, but may lead to bugs in the future.

My concern is not specific to this particular code change, but more to `Resolved.getAllIndices()` API change. Sorry for the confusion, I should post my comment there. I would be OK to change `Set` to `List` or an array for a local variable as long as it does not cause API change.

In the future, I would consider changing to a lazy evaluation of underlying collections (switching to `Stream` or `Iterator`).

Reply:
> In the future, I would consider changing to a lazy evaluation of underlying collections (switching to Stream or Iterator).
 
Amen.

> My concern is not specific to this particular code change, but more to Resolved.getAllIndices() API change. Sorry for the confusion, I should post my comment there. I would be OK to change Set to List or an array for a local variable as long as it does not cause API change.

That I agree with. Indeed public API-level concerns are important. I’ll change that place to use Sets.

Reply:
Please rebase against master during the next round of changes to simplify PR review